### PR TITLE
Extract image

### PR DESCRIPTION
**Pull request by CluckRookie**

### CODE REVIEW

File: uniflow/op/model/layout_utils.py [hunk 1]
```diff
@@ -0,0 +1,348 @@
...
+        Predicts the layout of the input image.
+
+        Args:
+            img (numpy.ndarray): Input image.
+
+        Returns:
+            list: List of dictionaries containing the predicted layout information.
+                  Each dictionary contains the type of the element and its bounding box coordinates.
+        """
+        h, w = (1024, 768)
```
Comment (CambioML) [inline uniflow/op/model/layout_utils.py:324]: Let's move this out of the __call__ and use `H`, `W` to indicate this is class level member variable to improve readability.

File: uniflow/op/model/layout_utils.py [hunk 2]
```diff
@@ -0,0 +1,348 @@
...
+        img (numpy.ndarray): The input image.
+        input_size (tuple): The desired size of the input image.
+        swap (tuple, optional): The order of dimensions to swap. Defaults to (2, 0, 1).
+
+    Returns:
+        numpy.ndarray: The preprocessed image.
+        float: The scaling factor applied to the image.
+    """
+    if len(img.shape) == 3:
+        padded_img = np.ones((input_size[0], input_size[1], 3), dtype=np.uint8) * 114
```
Comment (CambioML) [inline uniflow/op/model/layout_utils.py:24]: qq: curious about why `114` is the default padding value?

File: uniflow/op/model/layout_utils.py [hunk 3]
```diff
@@ -0,0 +1,348 @@
...
+    r = min(input_size[0] / img.shape[0], input_size[1] / img.shape[1])
+    resized_img = cv2.resize(
+        img,
+        (int(img.shape[1] * r), int(img.shape[0] * r)),
+        interpolation=cv2.INTER_LINEAR,
+    ).astype(np.uint8)
+    padded_img[: int(img.shape[0] * r), : int(img.shape[1] * r)] = resized_img
+
+    padded_img = padded_img.transpose(swap)
+    padded_img = np.ascontiguousarray(padded_img, dtype=np.float32)
```
Comment (CambioML) [inline uniflow/op/model/layout_utils.py:37]: very interesting. I was not aware of this optimization. TIL!!!

File: uniflow/op/model/layout_utils.py [hunk 4]
```diff
@@ -0,0 +1,348 @@
+"""Layout Module."""
+from typing import List
+
+import cv2
+import numpy as np
+import onnxruntime
+from huggingface_hub import hf_hub_download
+
+
+def preprocess(img, input_size, swap=(2, 0, 1)):
```
Comment (CambioML) [inline uniflow/op/model/layout_utils.py:10]: nit: please add all missing typehint and docstring (argument, return, ...) for this Non Maximum Suppression part.

File: uniflow/op/model/layout_utils.py [hunk 5]
```diff
@@ -0,0 +1,348 @@
...
+        return None
+    valid_scores = cls_scores[valid_score_mask]
+    valid_boxes = boxes[valid_score_mask]
+    valid_cls_inds = cls_inds[valid_score_mask]
+    keep = nms(valid_boxes, valid_scores, nms_thr)
+    if keep:
+        dets = np.concatenate(
+            [valid_boxes[keep], valid_scores[keep, None], valid_cls_inds[keep, None]], 1
+        )
+    return dets
```
Comment (CambioML) [inline uniflow/op/model/layout_utils.py:120]: consider to group all these methods into a class with static method to improve readability.
class name like: `nmsprocess` with static methods
- nms_single_class
- nms_multiclass
- nms_multiclass_class_aware
- nms_multiclass_class_agnostic

File: uniflow/op/model/layout_utils.py (hunk 2 above)
Comment (CluckRookie) [reply uniflow/op/model/layout_utils.py:24]: yolox default padding value is 114
https://github.com/Megvii-BaseDetection/YOLOX/blob/main/demo/OpenVINO/cpp/yolox_openvino.cpp#L36

**Review by CambioML — APPROVED** [review]

**Pull request merged**